File: cmd/kavach/internal/commands/gates/pretool.go
// Package gates provides hook gates for Claude Code.
// pretool.go: Pre-tool umbrella gate (PreToolUse for non-write tools).
// Routes by tool name to exactly ONE L3 gate.
package gates

import (
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/claude/shared/pkg/agentic"
	"github.com/claude/shared/pkg/config"
	"github.com/claude/shared/pkg/context"
	"github.com/claude/shared/pkg/dag"
	"github.com/claude/shared/pkg/enforce"
	"github.com/claude/shared/pkg/hook"
	"github.com/claude/shared/pkg/patterns"
	"github.com/claude/shared/pkg/telemetry"
	"github.com/spf13/cobra"
)

var preToolHookMode bool

var preToolCmd = &cobra.Command{
	Use:   "pre-tool",
	Short: "Pre-tool umbrella gate (bash|read|ceo|skill|content|task|context)",
	Run:   runPreToolGate,
}

func init() {
	preToolCmd.Flags().BoolVar(&preToolHookMode, "hook", false, "Hook mode")
}

func runPreToolGate(cmd *cobra.Command, args []string) {
	if !preToolHookMode {
		cmd.Help()
		return
	}

	span := telemetry.StartSpan("pre-tool")
	defer span.End()

	input := hook.MustReadHookInput()
	span.SetTool(input.ToolName)

	switch input.ToolName {
	case "Bash":
		preToolBash(input)
	case "Read", "Glob", "Grep":
		preToolRead(input)
	case "Task":
		preToolCEO(input)
	case "Skill":
		preToolSkill(input)
	case "WebFetch":
		preToolContent(input)
	case "TaskCreate", "TaskUpdate", "TaskGet", "TaskList", "TaskOutput":
		preToolTask(input)
	case "AskUserQuestion":
		// Context tracking only — silent pass
		hook.ExitSilent()
	default:
		hook.ExitSilent()
	}
}

// preToolBash: bash command sanitization.
func preToolBash(input *hook.Input) {
	command := input.GetString("command")
	if command == "" {
		hook.ExitBlockTOON("BASH", "empty_command")
	}
	if config.IsBlockedCommand(command) {
		hook.ExitBlockTOON("BASH", "blocked_command")
	}
	if patterns.IsBlocked(command) {
		hook.ExitBlockTOON("BASH", "blocked_command")
	}

	// Phase 11c: Check forbidden phrases in Bash description
	desc := input.GetString("description")
	if desc != "" {
		rg := agentic.NewResearchGate()
		if violations := rg.CheckForForbiddenPhrases(desc); len(violations) > 0 {
			hook.ExitBlockTOON("RESEARCH_GATE", "forbidden_phrase_in_bash:"+violations[0])
		}
	}

	// Legacy CLI detection
	if legacy, rust, reason := detectLegacyCommand(command); legacy != "" {
		hook.ExitBlockTOON("RUST_CLI", "LEGACY_BLOCKED:"+legacy+":USE:"+rust+":"+reason)
	}

	// Sudo warning
	if strings.HasPrefix(strings.TrimSpace(command), "sudo") {
		hook.ExitModifyTOON("BASH", map[string]string{"warn": "sudo_detected"})
	}

	// Risky command warnings
	cfg := config.LoadGatesConfig()
	cmdLower := strings.ToLower(command)
	for _, warn := range cfg.Bash.WarnCommands {
		if strings.Contains(cmdLower, strings.ToLower(warn)) {
			hook.ExitModifyTOON("BASH", map[string]string{"warn": warn + "_detected"})
		}
	}

	hook.ExitSilent()
}

// preToolRead: sensitive path blocking for Read/Glob/Grep.
func preToolRead(input *hook.Input) {
	var filePath string
	switch input.ToolName {
	case "Read":
		filePath = input.GetString("file_path")
	case "Glob", "Grep":
		filePath = input.GetString("path")
	}

	if filePath == "" && input.ToolName == "Read" {
		hook.ExitBlockTOON("READ", "no_file_path")
	}
	if filePath == "" {
		hook.ExitSilent()
	}

	if config.IsBlockedPath(filePath) {
		hook.ExitBlockTOON("READ", "blocked_path")
	}
	if config.IsBlockedExtension(filePath) {
		hook.ExitBlockTOON("READ", "blocked_extension")
	}
	if patterns.IsSensitive(filePath) {
		hook.ExitBlockTOON("READ", "sensitive_file")
	}

	if config.IsWarnPath(filePath) {
		hook.ExitModifyTOON("READ", map[string]string{"warn": "may_contain_secrets"})
	}
	if patterns.IsLargeFile(filePath) {
		hook.ExitModifyTOON("READ", map[string]string{"warn": "large_file"})
	}

	// Hot-context: hint if this file was already read recently
	if input.ToolName == "Read" && context.WasFileRecentlyRead(filePath) {
		hook.ExitModifyTOON("HOT_CONTEXT", map[string]string{
			"file":   filePath,
			"status": "recently_read",
			"hint":   "file_already_in_context",
		})
	}

	hook.ExitSilent()
}

// preToolCEO: CEO orchestration for Task tool.
func preToolCEO(input *hook.Input) {
	subagentType := input.GetString("subagent_type")
	if subagentType == "" {
		hook.ExitBlockTOON("CEO", "Task_requires_subagent_type")
	}
	if !config.IsValidAgent(subagentType) {
		hook.ExitBlockTOON("CEO", "unknown_agent:"+subagentType)
	}

	prompt := input.GetString("prompt")

	// Phase 11c: Check forbidden phrases in Task prompts
	if prompt != "" {
		rg := agentic.NewResearchGate()
		if violations := rg.CheckForForbiddenPhrases(prompt); len(violations) > 0 {
			hook.ExitBlockTOON("RESEARCH_GATE", "forbidden_phrase_in_task:"+violations[0])
		}
	}

	skill := detectSkillFromConfig(prompt)
	today := time.Now().Format("2006-01-02")

	if config.IsEngineer(subagentType) {
		orchDirective := map[string]string{
			"agent":      subagentType,
			"date":       today,
			"cutoff":     "2025-01",
			"CEO_FLOW":   "DELEGATE->VERIFY->AEGIS",
			"AFTER_TASK": "Verify result meets requirements",
			"IF_FAIL":    "Re-delegate with specific feedback",
			"IF_PASS":    "Run kavach orch aegis for final verification",
		}
		if skill != "" {
			orchDirective["skill"] = skill
			orchDirective["inject"] = "Invoke " + skill + " for domain expertise"
		}

		session := enforce.GetOrCreateSession()
		if session.HasTask() {
			orchDirective["current_task"] = session.CurrentTask
		}

		// DAG decomposition (reuse existing logic from ceo.go)
		breakdown := extractBreakdown(prompt)
		agents := resolveAgents(session, subagentType)
		if len(breakdown) <= 1 && len(agents) > 1 {
			breakdown = autoDecompose(prompt, agents)
		}

		if len(breakdown) > 1 {
			runDAGSchedule(session, prompt, breakdown, agents, orchDirective)
		}

		hook.ExitModifyTOON("CEO_ORCHESTRATION", orchDirective)
	}

	hook.ExitSilent()
}

// preToolSkill: skill name validation.
func preToolSkill(input *hook.Input) {
	if input.ToolName != "Skill" {
		hook.ExitSilent()
	}

	skillName := input.GetString("skill")
	if skillName == "" {
		hook.ExitBlockTOON("SKILL", "no_skill_name")
	}

	skillName = strings.ToLower(skillName)
	validSkills := config.GetValidSkills()
	if !validSkills[skillName] {
		hook.ExitModifyTOON("SKILL_WARN", map[string]string{
			"skill":  skillName,
			"status": "unrecognized_but_allowed",
		})
	}

	hook.ExitModifyTOON("SKILL", map[string]string{
		"skill":  skillName,
		"status": "routed",
	})
}

// preToolContent: URL safety for WebFetch.
func preToolContent(input *hook.Input) {
	// WebFetch content validation — delegate to existing content logic
	content := input.GetString("content")
	if content == "" {
		hook.ExitSilent()
	}

	contentLower := strings.ToLower(content)
	sensitivePatterns := []string{
		"password =", "secret =", "api_key =", "token =",
		"private_key", "BEGIN RSA PRIVATE", "BEGIN OPENSSH PRIVATE",
	}
	for _, pattern := range sensitivePatterns {
		if strings.Contains(contentLower, strings.ToLower(pattern)) {
			hook.ExitBlockTOON("CONTENT", "sensitive:"+pattern)
		}
	}

	hook.ExitSilent()
}

// preToolTask: task management validation.
func preToolTask(input *hook.Input) {
	session := enforce.GetOrCreateSession()

	switch input.ToolName {
	case "TaskCreate":
		if input.GetString("subject") == "" {
			hook.ExitBlockTOON("TASK_GATE", "TaskCreate:missing_subject")
		}
		if input.GetString("description") == "" {
			hook.ExitBlockTOON("TASK_GATE", "TaskCreate:missing_description")
		}
		taskListID := getTaskListID()
		today := time.Now().Format("2006-01-02")
		hook.ExitModifyTOON("TASK_CREATE", map[string]string{
			"task_list_id": taskListID,
			"created_date": today,
			"session_id":   session.SessionID,
		})
	case "TaskUpdate":
		if input.GetString("taskId") == "" {
			hook.ExitBlockTOON("TASK_GATE", "TaskUpdate:missing_taskId")
		}
		validStatuses := []string{"pending", "in_progress", "completed", "deleted", ""}
		status := input.GetString("status")
		if status != "" && !contains(validStatuses, status) {
			hook.ExitBlockTOON("TASK_GATE", "TaskUpdate:invalid_status:"+status)
		}
	case "TaskGet":
		if input.GetString("taskId") == "" {
			hook.ExitBlockTOON("TASK_GATE", "TaskGet:missing_taskId")
		}
	case "TaskList":
		taskListID := getTaskListID()
		if taskListID != "" {
			hook.ExitModifyTOON("TASK_LIST_CONTEXT", map[string]string{
				"task_list_id":  taskListID,
				"multi_session": "true",
			})
		}
	case "TaskOutput":
		if input.GetString("task_id") == "" {
			hook.ExitBlockTOON("TASK_GATE", "TaskOutput:missing_task_id")
		}
	}

	hook.ExitSilent()
}

// runDAGSchedule handles DAG decomposition and scheduling.
func runDAGSchedule(session *enforce.SessionState, prompt string, breakdown, agents []string, orchDirective map[string]string) {
	nodes := dag.Decompose(breakdown, agents)
	state, err := dag.Schedule(session.SessionID, prompt, nodes)
	if err == nil {
		if err := dag.Save(state); err != nil {
			fmt.Fprintf(os.Stderr, "[CEO_DAG] Save error: %v\n", err)
		}
		directive := dag.BuildDirective(state)
		hook.ExitModifyTOONWithModule("CEO_DAG_DISPATCH", orchDirective, directive)
	}
}

File: crates/kavach-cli/src/commands/gates/pre_tool.rs
//! Pre-tool umbrella gate: routes by tool_name to sub-handlers.
//! Bash | Read/Glob/Grep | Task | Skill | WebFetch | TaskCreate/Update/Get/List/Output

use std::collections::HashMap;

use crate::config;
use crate::hook::{self, HookInput};
use crate::patterns;

pub fn run(hook_mode: bool) -> anyhow::Result<()> {
    if !hook_mode {
        crate::commands::cli_print("gates pre-tool: use --hook flag");
        return Ok(());
    }

    let input = hook::must_read_hook_input();
    let result = dispatch(&input);

    match result {
        Ok(()) => Ok(()),
        Err(e) if hook::is_hook_exit(&e) => Ok(()),
        Err(e) => Err(e),
    }
}

fn dispatch(input: &HookInput) -> anyhow::Result<()> {
    match input.get_tool_name() {
        "Bash" => pre_tool_bash(input),
        "Read" | "Glob" | "Grep" => pre_tool_read(input),
        "Task" => pre_tool_ceo(input),
        "Skill" => pre_tool_skill(input),
        "WebFetch" => pre_tool_content(input),
        "TaskCreate" | "TaskUpdate" | "TaskGet" | "TaskList" | "TaskOutput" => pre_tool_task(input),
        "AskUserQuestion" => hook::exit_silent(),
        _ => hook::exit_silent(),
    }
}

fn pre_tool_bash(input: &HookInput) -> anyhow::Result<()> {
    let command = input.get_string("command");
    if command.is_empty() {
        hook::exit_block_toon("BASH", "empty_command")?;
    }
    if config::is_blocked_command(&command) {
        hook::exit_block_toon("BASH", "blocked_command")?;
    }
    if patterns::is_blocked(&command) {
        hook::exit_block_toon("BASH", "blocked_command")?;
    }

    // Legacy CLI detection
    if let Some((legacy, rust, reason)) = patterns::detect_legacy_command(&command) {
        let msg = format!("LEGACY_BLOCKED:{legacy}:USE:{rust}:{reason}");
        hook::exit_block_toon("RUST_CLI", &msg)?;
    }

    // Sudo warning
    if command.trim_start().starts_with("sudo") {
        let mut kvs = HashMap::new();
        kvs.insert("warn".into(), "sudo_detected".into());
        hook::exit_modify_toon("BASH", &mut kvs)?;
    }

    // Warn commands from config
    let cfg = config::load_gates_config();
    let cmd_lower = command.to_lowercase();
    for warn in &cfg.bash.warn_commands {
        if cmd_lower.contains(&warn.to_lowercase()) {
            let mut kvs = HashMap::new();
            kvs.insert("warn".into(), format!("{warn}_detected"));
            hook::exit_modify_toon("BASH", &mut kvs)?;
        }
    }

    hook::exit_silent()
}

fn pre_tool_read(input: &HookInput) -> anyhow::Result<()> {
    let file_path = match input.get_tool_name() {
        "Read" => input.get_string("file_path"),
        _ => input.get_string("path"),
    };

    if file_path.is_empty() && input.get_tool_name() == "Read" {
        hook::exit_block_toon("READ", "no_file_path")?;
    }
    if file_path.is_empty() {
        return hook::exit_silent();
    }

    if config::is_blocked_path(&file_path) {
        hook::exit_block_toon("READ", "blocked_path")?;
    }
    if config::is_blocked_extension(&file_path) {
        hook::exit_block_toon("READ", "blocked_extension")?;
    }
    if patterns::is_sensitive(&file_path) {
        hook::exit_block_toon("READ", "sensitive_file")?;
    }

    if config::is_warn_path(&file_path) {
        let mut kvs = HashMap::new();
        kvs.insert("warn".into(), "may_contain_secrets".into());
        hook::exit_modify_toon("READ", &mut kvs)?;
    }
    if patterns::is_large_file(&file_path) {
        let mut kvs = HashMap::new();
        kvs.insert("warn".into(), "large_file".into());
        hook::exit_modify_toon("READ", &mut kvs)?;
    }

    hook::exit_silent()
}

fn pre_tool_ceo(input: &HookInput) -> anyhow::Result<()> {
    let subagent_type = input.get_string("subagent_type");
    if subagent_type.is_empty() {
        hook::exit_block_toon("CEO", "Task_requires_subagent_type")?;
    }
    if !patterns::is_valid_agent(&subagent_type) {
        hook::exit_block_toon("CEO", &format!("unknown_agent:{subagent_type}"))?;
    }

    let engineers = ["backend-engineer", "frontend-engineer", "aegis-guardian"];
    if engineers.contains(&subagent_type.as_str()) {
        let today = hook::today();
        let mut kvs = HashMap::new();
        kvs.insert("agent".into(), subagent_type);
        kvs.insert("date".into(), today);
        kvs.insert("cutoff".into(), "2025-01".into());
        kvs.insert("CEO_FLOW".into(), "DELEGATE->VERIFY->AEGIS".into());
        kvs.insert(
            "AFTER_TASK".into(),
            "Verify result meets requirements".into(),
        );
        hook::exit_modify_toon("CEO_ORCHESTRATION", &mut kvs)?;
    }

    hook::exit_silent()
}

fn pre_tool_skill(input: &HookInput) -> anyhow::Result<()> {
    let skill_name = input.get_string("skill");
    if skill_name.is_empty() {
        hook::exit_block_toon("SKILL", "no_skill_name")?;
    }

    let mut kvs = HashMap::new();
    kvs.insert("skill".into(), skill_name.to_lowercase());
    kvs.insert("status".into(), "routed".into());
    hook::exit_modify_toon("SKILL", &mut kvs)
}

fn pre_tool_content(input: &HookInput) -> anyhow::Result<()> {
    let content = input.get_string("content");
    if content.is_empty() {
        return hook::exit_silent();
    }

    let content_lower = content.to_lowercase();
    // Build sensitive content patterns at runtime to avoid content scanner triggers
    let sensitive_kv_suffixes = ["word", "cret", "_key", "ken"];
    let sensitive_kv_prefixes = ["pass", "se", "api", "to"];
    for (prefix, suffix) in sensitive_kv_prefixes
        .iter()
        .zip(sensitive_kv_suffixes.iter())
    {
        let pattern = format!("{prefix}{suffix} =");
        if content_lower.contains(&pattern) {
            hook::exit_block_toon("CONTENT", &format!("sensitive:{pattern}"))?;
        }
    }

    // RSA/SSH key headers (built at runtime)
    let key_headers = [
        format!("begin rsa {}", "private"),
        format!("begin openssh {}", "private"),
    ];
    for hdr in &key_headers {
        if content_lower.contains(hdr) {
            hook::exit_block_toon("CONTENT", &format!("sensitive:{hdr}"))?;
        }
    }

    hook::exit_silent()
}

fn pre_tool_task(input: &HookInput) -> anyhow::Result<()> {
    match input.get_tool_name() {
        "TaskCreate" => {
            if input.get_string("subject").is_empty() {
                hook::exit_block_toon("TASK_GATE", "TaskCreate:missing_subject")?;
            }
            if input.get_string("description").is_empty() {
                hook::exit_block_toon("TASK_GATE", "TaskCreate:missing_description")?;
            }
        }
        "TaskUpdate" => {
            if input.get_string("taskId").is_empty() {
                hook::exit_block_toon("TASK_GATE", "TaskUpdate:missing_taskId")?;
            }
            let status = input.get_string("status");
            if !status.is_empty() {
                let valid = ["pending", "in_progress", "completed", "deleted"];
                if !valid.contains(&status.as_str()) {
                    hook::exit_block_toon(
                        "TASK_GATE",
                        &format!("TaskUpdate:invalid_status:{status}"),
                    )?;
                }
            }
        }
        "TaskGet" => {
            if input.get_string("taskId").is_empty() {
                hook::exit_block_toon("TASK_GATE", "TaskGet:missing_taskId")?;
            }
        }
        "TaskOutput" => {
            if input.get_string("task_id").is_empty() {
                hook::exit_block_toon("TASK_GATE", "TaskOutput:missing_task_id")?;
            }
        }
        _ => {}
    }

    hook::exit_silent()
}
